pkg/apis/kops/instancegroup.go

```diff
@@ -44,30 +44,71 @@ type InstanceGroup struct {
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
+// InstanceGroupList is a list of instance groups
 type InstanceGroupList struct {
 	metav1.TypeMeta `json:",inline"`
 	metav1.ListMeta `json:"metadata,omitempty"`
-
+	// Items is a collection of instancegroups
 	Items []InstanceGroup `json:"items"`
 }
 
 // InstanceGroupRole string describes the roles of the nodes in this InstanceGroup (master or nodes)
 type InstanceGroupRole string
 
 const (
-	InstanceGroupRoleMaster  InstanceGroupRole = "Master"
-	InstanceGroupRoleNode    InstanceGroupRole = "Node"
+	// InstanceGroupRoleMaster indicates a master node
+	InstanceGroupRoleMaster InstanceGroupRole = "Master"
+	// InstanceGroupRoleNode indicates a compute node
+	InstanceGroupRoleNode InstanceGroupRole = "Node"
+	// InstanceGroupRoleBastion indicates a bastion node
 	InstanceGroupRoleBastion InstanceGroupRole = "Bastion"
 )
 
+// AllInstanceGroupRoles is a collection of roles a node may be
 var AllInstanceGroupRoles = []InstanceGroupRole{
 	InstanceGroupRoleNode,
 	InstanceGroupRoleMaster,
 	InstanceGroupRoleBastion,
 }
 
+// RolloutStrategy defines the strategy to use when performing rollouts on this instance group
+type RolloutStrategy string
+
+// DuplicatStrategy contains options for the duplication rollout
+type DuplicatStrategy struct {
+	// DuplicateInstanceGroup indicate we should copy the entire group
+	DuplicatInstanceGroup bool `json:"duplicateInstanceGroup,omitempty"`
+}
+
+const (
+	// DefaultRollout indicates the default one by one with a time interval and or drain
+	DefaultRollout = "default"
+	// DuplicateRollout indicates a duplication of instancegroup
+	DuplicateRollout = "duplicate"
+	// ScaleUpRollout indicates a scalled ASG rollout
+	ScaleUpRollout = "scale-up"
+)
+
+// UpdateStrategy provides details about the rollout stratergy for a instancegroup
+type UpdateStrategy struct {
+	// Batch is a batch size to operate within
+	Batch int `json:"batch,omitempty"`
+	// Drain indicates if this group should be drained
+	Drain bool `json:"drain,omitempty"`
+	// DrainTimeout is the amount of time we wait for drain pods
+	DrainTimeout *metav1.Duration `json:"drainTimeout,omitempty"`
+	// Interval is the time to given between iterations
+	Interval *metav1.Duration `json:"interval,omitempty"`
+	// Name defines the strategy to use when performing a rollout on this instance group
+	Name RolloutStrategy `json:"name,omitempty"`
+	// PostDrainDelay is the duration we wait after draining each node
+	PostDrainDelay *metav1.Duration `json:"postDrainDelay,omitempty"`
+}
+
 // InstanceGroupSpec is the specification for a instanceGroup
 type InstanceGroupSpec struct {
+	// Strategy is the strategy to use for this instances group
+	Strategy *UpdateStrategy `json:"strategy,omitempty"`
```

Comment:
Interesting - I like the idea of putting this into the InstanceGroup, but it's going to make the logic tricky for the CLI - does the CLI override these options?

Reply:
Yes, the CLI overrides the instancegroup definition